---
title: "WTJR Genotype Environment Correlations"
output:
  rmdformats::html_clean:
    fig_width: 10
    thumbnails: FALSE
    lightbox: FALSE
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r, echo=F, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(warning = F, message = F, fig.align = "center", cache = F)
options(scipen = 1)

# Load packages -----------------------------------------------------------
library(rmdformats)
library(readxl)
library(raster)
library(GGally)
library(broom)
library(factoextra)
library(reactable)
library(cowplot)
library(rnaturalearth)
library(rnaturalearthhires)
library(geosphere)
library(raster)
library(cowplot)
library(corrplot)
library(tidyverse)
```


```{r load map data}
state_prov <- ne_states(c("united states of america", "canada"), returnclass = "sf")
countries <- ne_countries(continent = "north america", scale = 10, returnclass = "sf")
coast <- ne_coastline(scale = 10, returnclass = "sf")
```

A summary of genotype-environment correlations at the identified winter color loci in white-tailed jackrabbits.

## Samples used


```{r load sample info}
# Get samples from the bamlist
seqed_samples <- read_delim(file = "data/all_wtjr_sample_bamlist.txt", delim = "\t", col_names = "bam_file") %>% 
  mutate(sample = basename(bam_file)) %>% 
  mutate(sample = str_remove(sample, ".realigned.sorted.bam")) %>% 
  mutate(sample = str_remove(sample, "_realigned_sorted.bam"))


# My data
tim <- read_csv("data/labwork/possible_samples_to_extracts.csv") %>% 
  select(sample = ID_museum, year, latitude, longitude, winter_pheno, state, source = source.x) %>% 
  mutate(museum = year < 2008) %>% 
  mutate(museum = ifelse(is.na(museum), T, museum)) %>% 
  mutate(batch = "tim")

# Fix the LTW_TWO_4140 sample
tim$sample[tim$sample == "Mafalda_LTW_WYO_4140"] <- "LTW_COL_4140"
tim$year[tim$sample == "LTW_COL_4140"] <- NA
tim$latitude[tim$sample == "LTW_COL_4140"] <- 39.04814
tim$longitude[tim$sample == "LTW_COL_4140"] <- -105.56139
tim$winter_pheno[tim$sample == "LTW_COL_4140"] <- NA
tim$state[tim$sample == "LTW_COL_4140"] <- "CO"
tim$source[tim$sample == "LTW_COL_4140"] <- "LTW_COL_4140"
                                                

# Mafalda's data
mafalda <- read_excel("data/mf_thesis_table_S3_3.xlsx") %>% 
  filter(Species == "Lepus townsendii") %>%
  mutate(museum = ifelse(Origin == "Field", F, T),
         year = NA, 
         winter_pheno = NA,
         sample = str_replace(`Sample ID`, "_", "-"),
         batch = "mafalda") %>% 
  separate(`Location (Population)`, into = c("country", "state"), sep = "/") %>% 
  select(sample, year, latitude = `Lat.`, longitude = `Long.`,  winter_pheno, source = Origin, state, museum, batch) %>% 
  bind_rows(., read_csv("data/mafalda_UCM_samples.csv"))


wtjr <- seqed_samples %>% 
  left_join(bind_rows(tim, mafalda)) 


# wtjr %>% 
#   filter(batch == "tim") %>% 
#   write_csv("~/Desktop/tmp_TJT_samples_for_SOM.csv")
# 
# x <- wtjr %>% 
#   filter(batch == "mafalda")
```





I collected (to the best of my knowledge) all the raw reads data we have from sequenced WTJRs. This includes 82 samples from Mafalda's work (sequenced mostly to low coverage, but with a couple medium-coverage samples in there) and 62 samples from my work. The samples I sequenced had all been obtained by Mafalda: either from museums or sent by hunters or other folks. I obtained the location data for all these samples from Mafalda's "Megadatabase" excel spreadsheet.

Here is a map of the locations of these 143 samples:

```{r}
ggplot() +
  geom_sf(data = state_prov, color = "grey30", fill = "grey90") +
  geom_point(data = wtjr, aes(x = longitude, y = latitude, color = batch), size = 2) +
  geom_sf(data = countries, color = "grey10", fill = NA) +
  geom_sf(data= coast, color = "grey10", fill = NA) +
  coord_sf(
    xlim = c(-125, -92),
    ylim = c(32, 51),
    clip = "on", 
    expand = F) +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_rect(fill = NA))
```

So we have samples spanning the across our predicted range, though some sections are unrepresented: for my work we were more focused on sampling environmental breadth (in terms of bioclim variables) than geography itself. For each of our 142 samples, we can extract the environmental data for the three variables we use to predict color: snow cover, bioclim 2, and bioclim 3 (**N.B.**- Some samples come from the same or similar geographic coordinates, such that we do not have 142 unique environmental values). We can summarize the variability of these environmental metrics:  

```{r load environmental data}
env_variables <- raster::stack("../wtjr_sdm/processed_data/pheno_predictors_millsetal2018.tif")
names(env_variables) <- c("snow.cover", "bio_2", "bio_3")
```


```{r extract env data and combine}
env_preds <- raster::extract(env_variables, 
                             y = sp::SpatialPoints(data.frame(x = as.numeric(wtjr$longitude),
                                                              y = as.numeric(wtjr$latitude)), 
                                                   proj4string = CRS(proj4string(env_variables)))) %>% 
  as.data.frame(.)

wtjr_env <- wtjr %>% 
  bind_cols(env_preds)
```


```{r env summary table}
wtjr_env %>% 
  pivot_longer(snow.cover:bio_3, names_to = "env_variable", values_to = "value") %>% 
  group_by(env_variable) %>% 
  summarize(mean = mean(value),
            sd = sd(value),
            CV = sd/mean,
            min = min(value),
            max = max(value)) %>% 
  reactable(data = ., defaultColDef = colDef(
    cell = function(value) format(value, digits = 2, nsmall = 2),
    align = "center",
    minWidth = 70,
    headerStyle = list(background = "#71a1f0")
  ))
```

So, all three metrics are decently variable across our samples, with snow cover the most variable. 

## Bioinformatics pipeline

I can send along all the snakemake pipelines and other code I used, but here's a text description of the analysis pipeline.

I processed all reads through the same reads-to-bams bioinformatics pipeline: read trimming with `fastp`, mapping against the `DMNS18807_06042020_pseudohap2.1_10k.fasta` reference genome with `bwa`, duplicate removal with `picard MarkDuplicates`, and local indel realignment with `GATK`. 

From these mapped, filtered, and de-dupped `.bam` files, I used `ANGSD` to calculate genotype likelihoods, keeping only sites with at which we could get data for 50% of individuals (71 samples), along with a few other filters (mapping and base quality of 20, unique mappings with proper pairs only). Then, I used `PCAngsd` with default settings (which imposes a 5% minor allele frequency cutoff) to calculate genotype dosages, corrected for genome-wide population structure, at all SNPs that passed the above filters (this ends up being 25 million SNPs total). 

I then extracted these structure-corrected genotype dosages for all the associated SNPs listed in Table S6. I had a couple sites for which PCAangsd didn't output dosages for the given site, and I also had one site where the two alleles output by ANGSD didn't correspond to the alleles listed in the MASSArray table (Table S3). I ignored those sites. PCANGSD outputs dosages for the minor allele: in cases where the minor allele was the brown allele, a flipped the dosage to the opposite allele ($2-dosage$) so that all **dosages are for the white-associated allele**.

## Predictions

Since we're looking at the dosage of the white allele, we can make some *a priori* predictions, based both on ecology and on the effects of the environmental variables in our GLM for predicting coat color (Table S1 in the SOM, reproduced below):


```{r}
read.csv("~/Documents/work/postdoc_montana/projects/wtjr_sdm/results/pheno/glm_table_current_snow_cover.csv") %>% 
    reactable(data = ., defaultColDef = colDef(
    header = function(value) str_replace(value, "_", "\n"),
    cell = function(value) format(value, digits = 4, nsmall = 2),
    align = "center",
    minWidth = 90,
    headerStyle = list(background = "#71a1f0")
  ))
```

In this GLM, we're predicting the probability of winter white. So, increased snow cover has a positive effect on probability of winter white (as we'd expect ecologically). Increasing Bioclim 2 also has a positive effect on probability of white, while increasing Bioclim 3 has a negative effect on winter white. So, we should expect that genotype dosages should be correlated in the same direction: dosage of the white allele should be positively correlated with snow cover and Bioclim 2, and negatively correlated with Bioclim 3. I'll note here that the GLM we present above, and in Table S1, uses non-standardized predictors, so the effect sizes aren't comparable across samples. If we standardize, the ranking of environmental variables in terms of absolute value of effects is: snow cover (1.53), Bioclim 3 (-1.49), Bioclim 2 (0.88). 

## Sites to use

With dosages from multiple SNPs at each gene, we can look at dosage/environment correlations in a few different ways. In my original analysis, I looked at only the most-associated SNP for each gene. I repeat that analysis below ("top associated SNP"), but also try two other ways of looking at the genotype dosages. I looked at the average dosage of the white allele across all sites within a gene ("average dosage"), and I also selected the site with, on average, the most confident genotyping ("most informative"). For each sample and site, I calculated the Euclidean distance of the set of individual genotype likelihoods from [0.33, 0.33, 0.33]. The minimum distance (for a site with no data and uniform GLs) is thus 0, while for a site with known genotype (e.g., GLs of [0,1,0]), the maximum distance is `r sqrt(sum((c(0,1,0) - (1/3))^2))`. I rescaled the distances to the maximum, so that each genotype gets a "quality score" between 0 (no info) and 1 (definitive genotype). Then, for each gene, I selected the SNP with the highest average quality score. We can also try using these quality scores in whited correlations. 


```{r load GL quality scores}
asip_gl_scores <- read_csv("results/angsd/asip_GL_qc_scores.csv")
corin_gl_scores <- read_csv("results/angsd/corin_GL_qc_scores.csv")
ednrb_gl_scores <- read_csv("results/angsd/ednrb_GL_qc_scores.csv")
scaff380_gl_scores <- read_csv("results/angsd/scaff380_GL_qc_scores.csv")


asip_most_inform <- asip_gl_scores %>% 
  group_by(location) %>% 
  summarize(mean = mean(qc_score)) %>% 
  ungroup() %>% 
  slice_max(mean, n = 1)

corin_most_inform <- corin_gl_scores %>% 
  group_by(location) %>% 
  summarize(mean = mean(qc_score)) %>% 
  ungroup() %>% 
  slice_max(mean, n = 1)

ednrb_most_inform <- ednrb_gl_scores %>% 
  group_by(location) %>% 
  summarize(mean = mean(qc_score)) %>% 
  ungroup() %>% 
  slice_max(mean, n = 1)

scaff380_most_inform <- scaff380_gl_scores %>% 
  group_by(location) %>% 
  summarize(mean = mean(qc_score)) %>% 
  ungroup() %>% 
  slice_max(mean, n = 1)
```


```{r load genotype dosages}
read_in_white_dosage <- function(filename) {
  white_dosage <- read_csv(filename) %>% 
    select(-scaffold, -position) %>% 
    t(.) %>% 
    as_tibble(., rownames = "sample") 
  colnames(white_dosage) <- white_dosage[1,]
  colnames(white_dosage)[1] <- "sample"
  white_dosage <- white_dosage[-1,]
  
  # Return
  white_dosage %>% 
    mutate(across(-sample, as.numeric))
}

# All associated SNPs for which we have dosages,
# in terms of white dosage
corin_white_dosage <- read_in_white_dosage("results/pcangsd/corin_dosages_white.csv")
asip_white_dosage <- read_in_white_dosage("results/pcangsd/asip_dosages_white.csv")
ednrb_white_dosage <- read_in_white_dosage("results/pcangsd/ednrb_dosages_white.csv")
scaff380_white_dosage <- read_in_white_dosage("results/pcangsd/scaff380_dosages_white.csv")


# Get top associated SNP across each site
corin_top_white <- corin_white_dosage %>% 
  dplyr::select(sample, top = `342_46966737`) %>% 
  mutate(gene = "corin")

asip_top_white <- asip_white_dosage %>% 
  dplyr::select(sample, top = `245_24236852`) %>% 
  mutate(gene = "asip")

ednrb_top_white <- ednrb_white_dosage %>% 
  dplyr::select(sample, top = `311_3413288`) %>% 
  mutate(gene = "ednrb")

scaff380_top_white <- scaff380_white_dosage %>% 
  dplyr::select(sample, top = `380_35303819`) %>% 
  mutate(gene = "scaff380")

# Get mean white dosage across SNPs
corin_mean_white <- corin_white_dosage %>% 
  column_to_rownames("sample") %>% 
  rowMeans(., na.rm = T) %>% 
  as_tibble(., rownames = "sample") %>% 
  rename(avg = value) %>% 
  mutate(gene = "corin")


asip_mean_white <- asip_white_dosage %>% 
  column_to_rownames("sample") %>% 
  rowMeans(., na.rm = T) %>% 
  as_tibble(., rownames = "sample") %>% 
  rename(avg = value) %>% 
  mutate(gene = "asip")


ednrb_mean_white <- ednrb_white_dosage %>% 
  column_to_rownames("sample") %>% 
  rowMeans(., na.rm = T) %>% 
  as_tibble(., rownames = "sample") %>% 
  rename(avg = value) %>% 
  mutate(gene = "ednrb")

scaf380_mean_white <- scaff380_white_dosage %>% 
  column_to_rownames("sample") %>% 
  rowMeans(., na.rm = T) %>% 
  as_tibble(., rownames = "sample") %>% 
  rename(avg = value) %>% 
  mutate(gene = "scaff380")

# Get most informative locus across each SNP
corin_inform_white <- corin_white_dosage %>% 
  dplyr::select(sample, inform = `342_46871251`) %>% 
  mutate(gene = "corin")

asip_inform_white <- asip_white_dosage %>% 
  dplyr::select(sample, inform = `245_24204540`) %>% 
  mutate(gene = "asip")

ednrb_inform_white <- ednrb_white_dosage %>% 
  dplyr::select(sample, inform = `311_3402184`) %>% 
  mutate(gene = "ednrb")

scaff380_inform_white <- scaff380_white_dosage %>% 
  dplyr::select(sample, inform = `380_35306181`) %>% 
  mutate(gene = "scaff380")

```


```{r combine it all}
dosage_means <- bind_rows(asip_mean_white, corin_mean_white, ednrb_mean_white, scaf380_mean_white) %>% 
  select(sample, gene, avg)
dosage_top <- bind_rows(asip_top_white, corin_top_white, ednrb_top_white, scaff380_top_white)
dosage_inform <- bind_rows(asip_inform_white, corin_inform_white, ednrb_inform_white, scaff380_inform_white)

# Combine it all
wtjr_white_dosages <- wtjr_env %>% 
  left_join(dosage_means) %>% 
  left_join(dosage_top) %>% 
  left_join(dosage_inform)
```


## Correlation results

```{r corr plot fxn}
make_cor_plots <- function(wtjr_df) {
  for_top_cor <- wtjr_df %>% 
    dplyr::select(sample, snow.cover, bio_2, bio_3, gene, top) %>% 
    pivot_wider(names_from = gene, values_from = top) %>% 
    dplyr::select(-sample) %>% 
    relocate(asip:scaff380)
  
  for_avg_cor <- wtjr_df %>% 
    dplyr::select(sample, snow.cover, bio_2, bio_3, gene, avg) %>% 
    pivot_wider(names_from = gene, values_from = avg) %>% 
    dplyr::select(-sample) %>% 
    relocate(asip:scaff380)
  
  for_inform_cor <- wtjr_df %>% 
    dplyr::select(sample, snow.cover, bio_2, bio_3, gene, inform) %>% 
    pivot_wider(names_from = gene, values_from = inform) %>% 
    dplyr::select(-sample) %>% 
    relocate(asip:scaff380)
  
  # Top cor plot
  corr.mat <- cor(for_top_cor)
  pmat <- cor.mtest(for_top_cor, conf.level = 0.95)
  corr_plot <- corrplot(corr.mat, p.mat = pmat$p, sig.level = 0.05,
                        diag = F, method = 'circle', insig='blank', title = "\nTop associated SNPs")$corrPos
  text(corr_plot$x, corr_plot$y, round(corr_plot$corr, 2))
  
  # avg cor plot
  corr.mat <- cor(for_avg_cor)
  pmat <- cor.mtest(for_avg_cor, conf.level = 0.95)
  corr_plot <- corrplot(corr.mat, p.mat = pmat$p, sig.level = 0.05,
                        diag = F, method = 'circle', insig='blank', title = "\nAverage dosages")$corrPos
  text(corr_plot$x, corr_plot$y, round(corr_plot$corr, 2))
  
  # inform cor plot
  corr.mat <- cor(for_inform_cor)
  pmat <- cor.mtest(for_inform_cor, conf.level = 0.95)
  corr_plot <- corrplot(corr.mat, p.mat = pmat$p, sig.level = 0.05,
                        diag = F, method = 'circle', insig='blank', title = "\nMost informative SNP")$corrPos
  text(corr_plot$x, corr_plot$y, round(corr_plot$corr, 2))
  
  return(invisible(NULL))
}
```

### All samples

So, with all that set up, we can look at genotype/environment correlations.  We'll start by looking at all `r length(unique(wtjr_white_dosages$sample))` samples across the range. Below is a set of correlation plots, showing the strength of correlation between variables: cells without colored circles are not significantly correlated (though I list the Pearson correlation coefficient anyway, just to give the value). Cells with colored circles are significant (Pearson correlation test, p < 0.05). The plot titles give the details on which genotype dosages we're looking at: the top SNP, the average dosage, and the most informative SNP. 

```{r corr plot all sample}
wtjr_df <- wtjr_white_dosages
make_cor_plots(wtjr_white_dosages)
```

So, no matter the SNPs you look at, white genotype is strongly positively correlated with snow cover at all three genes. This is the same as it was before. And, same as before, what's going on at EDNRB and with bioclim 3 are not always what we'd expect: we expect negative correlations with Bioclim3, and only get them consistently with EDNRB. And, looking at Bioclim 2, correlations at ASIP and EDNRB are consistent across SNPs (at least in direction), but the correlation between CORIN and Bioclim 2 is dependent on the SNP(s) we're looking at. But, if we focus on the top associated SNP (a pretty defensible *a priori* choice), then the set of correlations is mostly as we expect. 

In our meeting, we talked about some of the ways we could subsample the data to look at how these correlations varied across space (in and outside of Colorado) and in terms of our prior inference (GWAS samples vs. non-GWAS samples). So, let's look at that:

### CO samples only 

```{r CO samples}
co_samples <- wtjr_white_dosages %>% 
  filter(state %in% c("CO", "Colorado"))
```

First, we can look at the `r length(unique(co_samples$sample))` samples from CO: 

```{r corr plot CO samples}
make_cor_plots(co_samples)
```

Here, the results across SNPs are very consistent: snow cover and Bioclim2 are strongly correlated in the expected direction, and Bioclim 3 is also mostly strongly correlated, and mostly in the unexpected direction. 


### Non-co samples 

```{r non-CO samples}
non_co_samples <- wtjr_white_dosages %>% 
  filter(!(state %in% c("CO", "Colorado")))
```

Next, we can look at the `r length(unique(non_co_samples$sample))` samples from outside of CO: 

```{r corr plot non-CO samples}
make_cor_plots(non_co_samples)
```

As before, we see that the genotype-environment correlations much weaker and mostly non-significant when we consider samples from outside of CO. Here, as with within CO, looking across different SNPs doesn't make a huge difference. 

### GWAS samples


```{r GWAS samples}
disambig <- read_xlsx("data/WTJR_74lowcovsamples_code_disamb.xlsx") 

gwas_samples <- wtjr_white_dosages %>% 
  filter(sample %in% str_replace(disambig$SampleID, "_", "-"))
```


Next, we can look only at the 74 samples that were used for doing GWAS (which are also all in CO): 

```{r corr plot GWAS samples}
make_cor_plots(gwas_samples)
```

So, this is the strongest set of correlations yet, and is also pretty consistent across the SNP sets. 

### Non-GWAS samples

```{r non-GWAS sample}
non_gwas_samples <- wtjr_white_dosages %>% 
  filter(!(sample %in% gwas_samples$sample))
         

non_gwas_in_co <- non_gwas_samples %>% 
  filter(state %in% c("CO", "Colorado")) %>% 
  pull(sample) %>% 
  unique(.) %>% 
  length(.)

```


Next, we can look at only non-GWAS samples. That is `r length(unique(non_gwas_samples$sample))` samples total, of which `r non_gwas_in_co` are in CO. 

```{r corr plot non GWAS samples}
make_cor_plots(non_gwas_samples)
```

Here, the patterns are pretty similar to the non-CO samples, but just a little bit stronger so that some correlations with Bioclim2 and Bioclim3 reach significance, when they didn't before. The story with snow cover is basically the same as outside CO: positive correlation that is too weak to be significant. 

### Non-GWAS samples in CO only

```{r non-GWAS CO sample}
non_gwas_samples <- wtjr_white_dosages %>% 
  filter(!(sample %in% gwas_samples$sample))
         

non_gwas_in_co <- non_gwas_samples %>% 
  filter(state %in% c("CO", "Colorado"))
```

Finally, the 21 CO samples that weren't included in the GWAS. 

```{r corr plot C) non GWAS samples}
make_cor_plots(non_gwas_in_co)
```

The fact that snow cover is significantly correlated with ASIP dosage is a bit of a surprise: with such small sample sizes, its interesting that anything is significant. Also interesting that CORIN and EDNRB genotype dosages are negatively correlated in the non-GWAS CO samples. However, I will note that the snow cover/ASIP correlation there is strongly attenuated in the weighted correlation (next section), so I wouldn't put too much stock in it. 

## Best of both worlds: weighted correlation?

One idea I had while working on this: perhaps we could get the best of both worlds by looking at the top associated SNPs, but doing correlations weighted by the confidence in each genotype dosage. The base R implementation of weighted correlations doesn't give p-values, only the correlation, and also isn't quickly compatible with the package I'm using to make the nice correlation plots. So I just calculated the correlations for some of our different subsets of samples, and in short: weighted correlations don't make much of a difference to the overall story. So I don't show them in the report but I can send them along if you're interested.


```{r weighted correlations, eval = F}
# Get the scores for only the top SNPs
top_scores <- bind_rows(mutate(asip_gl_scores, gene = "asip"),
                        mutate(corin_gl_scores, gene = "corin"),
                        mutate(ednrb_gl_scores, gene = "ednrb"),
                        mutate(scaff380_gl_scores, gene = "scaff380")) %>% 
  filter(location %in% c("342_46966737", "245_24236852", "311_3413288", "380_35303819")) %>% 
  select(-location)
  
# Combine it all
wtjr_white_dosages_weighted <- wtjr_env %>% 
  left_join(dosage_top) %>% 
  left_join(top_scores)


# Function that calculates it all
get_weighted_correlations <- function(wtjr_dosages_with_weight) {
  # Subset down into data frames I can use with the cov.wt function
asip_top_weights <- wtjr_dosages_with_weight %>% 
  filter(gene == "asip") 

corin_top_weights <- wtjr_dosages_with_weight %>% 
  filter(gene == "corin") 

ednrb_top_weights <- wtjr_dosages_with_weight %>% 
  filter(gene == "ednrb") 

scaff380_top_weights <- wtjr_dosages_with_weight %>% 
  filter(gene == "scaff380") 
  

# Get the weighted corrs we car about
asip_weighted_cor <- cov.wt(x = select(asip_top_weights,
                  snow.cover, bio_2, bio_3, top),
       wt = asip_top_weights$qc_score, 
       cor = T)$cor[4,1:3]

corin_weighted_cor <- cov.wt(x = select(corin_top_weights,
                  snow.cover, bio_2, bio_3, top),
       wt = corin_top_weights$qc_score, 
       cor = T)$cor[4,1:3]

ednrb_weighted_cor <- cov.wt(x = select(ednrb_top_weights,
                  snow.cover, bio_2, bio_3, top),
       wt = ednrb_top_weights$qc_score, 
       cor = T)$cor[4,1:3]

scaff380_weighted_cor <- cov.wt(x = select(scaff380_top_weights,
                  snow.cover, bio_2, bio_3, top),
       wt = scaff380_top_weights$qc_score, 
       cor = T)$cor[4,1:3]



t(rbind(asip_weighted_cor,
        corin_weighted_cor,
        ednrb_weighted_cor,
        scaff380_weighted_cor))

}


wtjr_white_dosages_weighted %>% 
  filter(sample %in% gwas_samples$sample) %>% 
  get_weighted_correlations(.)

wtjr_white_dosages_weighted %>% 
  filter(sample %in% non_gwas_samples$sample) %>% 
  get_weighted_correlations(.)


wtjr_white_dosages_weighted %>% 
  filter(sample %in% non_co_samples$sample) %>% 
  get_weighted_correlations(.)


wtjr_white_dosages_weighted %>% 
  filter(sample %in% non_gwas_in_co$sample) %>% 
  get_weighted_correlations(.)

```


## Conclusions

So, did this reanalysis, where we looked at other SNPs and also looked at different sets of samples, shed any new light? Frankly, not really. The strange patterns we with some correlations of EDNRB, and the unexpected direction of the Bioclim3 correlations, aren't really changed by looking at other SNPs. And looking at other sets of samples doesn't clarify things a ton either. The story is about the same: snow cover is positively correlated with the dosage of the white allele. For Bioclim 2, the results are mostly as expected though a little less clear. for Bioclim 3, the correlations more in the unexpected direction. These correlations between genotypes and environmental variables are primarily driven by the Colorado samples, and in particular by the ones that are included in the GWAS. This is at least partly a function of the fact that we have the most samples from CO. 

So, in terms of the overall goal of saying whether the genetic architecture of winter color pattern is likely to be similar across the range, I think the reanalysis hasn't really changed things. It seems pretty reasonable, given what we know about color pattern and the identities o these genes. But it isn't a slam-dunk: adding more samples from outside CO (and even from within CO, but outside the GWAS) tends to weaken the correlation between genotype dosage and environmental variables. But it doesn't make it disappear completely. 

## Update on scaffold 380

I added in the SNPs from the associated region on scaffold 380. I didn't write them into the results above, but you can see the SNPs from that region in all the correlation charts. It, like the other genes, doesn't really display the expected pattern of negative correlation with Bioclim 3. With snow cover, though, white allele dosage at this locus is pretty strongly correlated. Like the other genes, correlations are strongest among the GWAS and CO samples, and the correlations are weaker outside CO. That "drop-off" in correlation is larger than at the other SNPs: scaff380 dosage is basically uncorrelated with snow cover outside CO, while the other genes tend to keep a weak and nonsignificant correlation. Interestingly, scaffold 380 does have a significant positive association with snow cover among the 21 CO samples that weren't in the GWAS. Not totally sure what to make of this in terms of our theory that the association at scaffold 380 could be due to an environmental factor that co-varies with the environmental variables that drive winter color. I guess maybe this is what we'd expect, if this un-measured environmental variable was really correlated with snow? But a little tricky: perhaps it is also what we'd expect if it really was somehow all due to color. Need to think on it more. 


```{r}
# Save a top corr plot

  for_top_cor <- wtjr_white_dosages %>% 
    dplyr::select(sample, snow.cover, bio_2, bio_3, gene, top) %>% 
    pivot_wider(names_from = gene, values_from = top) %>% 
    dplyr::select(-sample) %>% 
    relocate(asip:scaff380)

png("results/prelim_GEA_corrplot.png", width = 800, height = 800)
  corr.mat <- cor(for_top_cor)
  pmat <- cor.mtest(for_top_cor, conf.level = 0.95)
  corr_plot <- corrplot(corr.mat, p.mat = pmat$p, sig.level = 0.05,
                        diag = F, method = 'circle', insig='blank', title = " ",
                        tl.col = "black",
                        col = COL2("PRGn"),
                        type = "lower")$corrPos
  text(corr_plot$x, corr_plot$y, round(corr_plot$corr, 2), )
  dev.off()


```



